binary/src/Pos/Binary/Class/Core.hs

```diff
@@ -659,3 +760,252 @@ instance (i ~ G.C, GSerialiseProd f) => GSerialiseSum (G.M1 i c f) where
     fieldsForCon _ _ = cborError "Bad constructor number"
     decodeSum      0 = G.M1 <$> gdecodeSeq
     decodeSum      _ = cborError "bad constructor number"
+
+
+----------------------------------------------------------------------------
+-- Size expressions
+----------------------------------------------------------------------------
+
+(.:) :: (c -> d) -> (a -> b -> c) -> (a -> b -> d)
+f .: g = \x y -> f (g x y)
+
+-- | Expressions describing the statically-computed size bounds on
+--   a type's possible values.
+type Size = Fix SizeF
+
+-- | The base functor for @Size@ expressions.
+data SizeF t
+    = AddF t t -- ^ Sum of two sizes.
+    | MulF t t -- ^ Product of two sizes.
+    | SubF t t -- ^ Difference of two sizes.
+    | AbsF t   -- ^ Absolute value of a size.
+    | NegF t   -- ^ Negation of a size.
+    | SgnF t   -- ^ Signum of a size.
+    | CasesF [Case t] -- ^ Case-selection for sizes. Used for sum types.
+    | ValueF Byte     -- ^ A constant value.
+    | ApF String (Byte -> Byte) t
+      -- ^ Application of a monotonic function to a size.
+    | forall a. Bi a => TodoF (forall x. Bi x => Proxy x -> Size) (Proxy a)
+      -- ^ A suspended size calculation ("thunk").
+  deriving Typeable
+
+instance Functor SizeF where
+    fmap f = \case
+        AddF x y  -> AddF (f x) (f y)
+        MulF x y  -> MulF (f x) (f y)
+        SubF x y  -> SubF (f x) (f y)
+        AbsF x    -> AbsF (f x)
+        NegF x    -> NegF (f x)
+        SgnF x    -> SgnF (f x)
+        CasesF xs -> CasesF (map (fmap f) xs)
+        ValueF x  -> ValueF x
+        ApF n g x -> ApF n g (f x)
+        TodoF g x -> TodoF g x
+
+instance Num (Fix SizeF) where
+    (+) = Fix .: AddF
+    (*) = Fix .: MulF
+    (-) = Fix .: SubF
+    negate = Fix . NegF
+    abs    = Fix . AbsF
+    signum = Fix . SgnF
+    fromInteger = Fix . ValueF . fromInteger
+
+instance Buildable t => Buildable (SizeF t) where
+    build x_ =
+        let showp2 c x y = bprint ("(" % build % " " % string % " " % build % ")") x c y
+        in case x_ of
+            AddF x y -> showp2 "+" x y
+            MulF x y -> showp2 "*" x y
+            SubF x y -> showp2 "-" x y
+            NegF x   -> bprint ("-" % build) x
+            AbsF x   -> bprint ("|" % build % "|") x
+            SgnF x   -> bprint ("sgn(" % build % ")") x
+            CasesF xs -> bprint ("{ " % build % "}") $ foldMap (bprint (build % " ")) xs
+            ValueF x  -> bprint shown (toInteger x)
+            ApF n _ x -> bprint (string % "(" % build % ")") n x
+            TodoF _ x -> bprint ("(_ :: " % shown % ")") (typeRep x)
+
+instance Buildable (Fix SizeF) where
+    build x = bprint build (unfix x)
+
+-- | Create a case expression from individual cases.
+szCases :: [Case Size] -> Size
+szCases = Fix . CasesF
+
+-- | An individual labeled case.
+data Case t = Case String t deriving (Typeable, Functor)
```

Comment:
Would it make sense to have an ADT instead of a string here, to give consumers of these sizes more of a hint what these cases represent? For instance, I think you use the strings `minBound` and `maxBound` in various places.. it  might be useful for consumers to _know_ that, so that they can derive best and worst cases (just as one example).

Reply:
This would be great, but slightly messy. The issue is that some cases are just named by purpose ("minBound" etc), while others are the actual constructor names used in a sum type (for example, "PubKeyASD", "ScriptASD", "RedeemASD" for `Bi AddrSpendingData`).

I could cover all of this with some ADT along the lines of

```haskell
data CaseLabel
    = MinBound
    | MaxBound
    | Constructor Con -- from TH?
```

but it isn't clear to me if the benefits outweigh the complexity. It still wouldn't let you do something like say "I'm only interested in the lower bound".